### PR TITLE
Generate a unique prefix instead of initials when naming subcharts

### LINKED ISSUE #262: Generate a unique prefix instead of initials when naming subcharts being pushed to the staging registry
We need to generate a unique prefix instead of using the initials when pushing charts to the staging registry. The original reason for doing this was to prevent overwriting the subcharts of one application with the subcharts of another. 

The other option is to use the chart version in which case overwriting wouldn't cause any damage as the two charts would (ideally) be the same 

__Originally posted by @nitishm in https://github.com/Azure/orkestra/pull/247#discussion_r635433686__

### PR DESCRIPTION
**Pull request by mahalrs**
Closes #262

This change generates a unique prefix (using SHA-256 hash of application name) instead of initials when naming subcharts.

### CODE REVIEW

**Comment by mahalrs** [comment]
@nitishm I'm not sure why e2e test is failing when I only replaced ToInitials() with GetHash(). Can you help me on this.

**Comment by nitishm** [comment]
@mahalrs this is the error when you run it locally using 
`make clean && make dev && make test`
```
lastTransitionTime: "2021-06-07T20:00:22Z"                                                                                                                                                               
     message: |-                                                                                                                                                                                              
       Helm install failed: ServiceAccount "bookinfo-a82a4dac39a740d62d77081e3a1c040d134e5249002422d84740307d486572bb-details" is invalid: metadata.labels: Invalid value: "a82a4dac39a740d62d77081e3a1c040d134e5249002422d84740307d486572bb-details": must be no more than 63 characters
```

We might need to truncate the hash to 2 or 3 chars at most to keep the names under the max len limit enforced by DNS1123 (i.e. 63 chars)

**Comment by mahalrs** [comment]
> @mahalrs this is the error when you run it locally using
> `make clean && make dev && make test`
> 
> ```
> lastTransitionTime: "2021-06-07T20:00:22Z"                                                                                                                                                               
>      message: |-                                                                                                                                                                                              
>        Helm install failed: ServiceAccount "bookinfo-a82a4dac39a740d62d77081e3a1c040d134e5249002422d84740307d486572bb-details" is invalid: metadata.labels: Invalid value: "a82a4dac39a740d62d77081e3a1c040d134e5249002422d84740307d486572bb-details": must be no more than 63 characters
> ```
> 
> We might need to truncate the hash to 2 or 3 chars at most to keep the names under the max len limit enforced by DNS1123 (i.e. 63 chars)

Should we use SHA-1 which is 40 characters.

**Comment by mahalrs** [comment]
@nitishm We will use the SHA-256 and then after appending the ```scc.Metadata.Name``` to it, we will truncate it from left side to make sure ```hash + scc.Metadata.Name``` <= 63 chars.

For example, ```a82a4dac39a740d62d77081e3a1c040d134e5249002422d84740307d486572bb-details``` will become ~~a82a4dac3~~```9a740d62d77081e3a1c040d134e5249002422d84740307d486572bb-details```.

Let me know what you think.

**Comment by nitishm** [comment]
That sounds pretty good. Add some test cases and also a boundary check where the app name itself is 63 chars. Though a corner case how would we handle that? Maybe truncate the name portion to accommodate part of the hash?

So it could be <partial-hash>-<truncated-name-to-accomodate-partial-hash>. But do this only for the corner case

**Comment by nitishm** [comment]
@jonathan-innis Any thoughts on this one?

**Comment by mahalrs** [comment]
@nitishm Max len limit 63 chars. So, at least 10 chars should come from hash, 1 char "-", and remaining from application name. So, max allowed application name is 52 chars. If > 52, truncate app name.

File: controllers/appgroup_reconciler.go [hunk 1]
```diff
@@ -147,7 +147,8 @@ func (r *ApplicationGroupReconciler) reconcileApplications(l logr.Logger, appGro
 					}
 				}
 
-				scc.Metadata.Name = utils.ConvertToDNS1123(utils.ToInitials(appCh.Metadata.Name) + "-" + scc.Metadata.Name)
+				hashAppChName := utils.GetHash(appCh.Metadata.Name)
```
Comment (jonathan-innis) [inline controllers/appgroup_reconciler.go:150]: Can we combine all these functions into a single `GetChartName` function? This will make things more readable

**Review by jonathan-innis — CHANGES_REQUESTED** [review]
Small readability comment, other than this looks great @mahalrs 🚀

File: pkg/utils/chart.go [hunk 2]
```diff
@@ -0,0 +1,17 @@
+package utils
+
+const (
+	DNS1123NameMaximumLength = 63
+
+	// subchartNameMaxLen is the maximum length of a chart/subchart name.
```
Comment (nitishm) [inline pkg/utils/chart.go:6]:
```suggestion
	// subchartNameMaxLen is the maximum length of the subchart name.
```

**Review by nitishm — APPROVED** [review]
LGTM 🚀

**Review by jonathan-innis — APPROVED** [review]
🚀
